### PR TITLE
#159987690 User should get verification email after signup

### PR DESCRIPTION
**Pull request by emmaadesile**
### What does this PR do?
- Implement the feature that sends a verification email to a user after successful signup.
- Confirms user's email after clicking on the verification link in the email.
- Resend email verification to user when they request for it.

### Description of Task to be completed?
- added two new endpoints to the routes
```
  POST /api/users/confirmation/:token
  POST /api/users/verify/resend-email
  ```
- user should get a verification email after successful signup
- user's email should be verified when they hit this route `/api/users/confirmation/:token`;
- user can also request that the verification email be resent when they make a post request to this route `/api/users/verify/resend-email`

### How should this be manually tested?
- After cloning the app
- create a `.env` file from `.envsample`
- run `npm install`
- Install and setup Postgres
- run `npm start:dev` to start the development server
- on postman, make a `POST` request to `/api/users` to signup a new user
- You should a message that a verification email has been sent to the user
- Copy the link from the email to make a `GET` request to `/api/users/confirmation/:token`
- You should get a message that the email has been confirmed
- Request that verification email be resent by making a `POST` request to  `/api/users/verify/resend-email`

### What are the relevant pivotal tracker stories?
[159987690](https://www.pivotaltracker.com/story/show/159987690)

### CODE REVIEW

File: server/controllers/emailVerificationController.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        return User
+          .update({ emailVerified: true }, {where: {id}})
+          .then(emailVerified => (
+            res.status(200).json({
+              status: 'success',
+              message: 'Email confirmed successfully. You can now login',
+            })
+          ))
+        }) 
+        .catch(error => console.log(error))
```
Comment (aknwosu) [inline server/controllers/emailVerificationController.js]: console.log? Properly catch errors sending appropriate message and status

**Comment by aknwosu** [comment]
@emmaadesile Sync with @CEOehis to fix Travis error, standardize your object notations. e.g. use { expect } or {expect} consistently with { expect } being the preferred

**Review by emekafredy — COMMENTED** [review]
Resolve issues with houndCI

File: server/controllers/emailVerificationController.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  static resendVerificationEmail(req, res) {
+    const { email } = req.body;
+    email.trim();
+
+    return User.findOne({ where: { email } })
+      .then((user) => {
+        if (!user) {
+          return res.status(404).json({
+            status: 'failed',
+            message: 'User does not exist in database'
```
Comment (aknwosu) [inline server/controllers/emailVerificationController.js:104]: message: 'User does not exist'

File: server/controllers/emailVerificationController.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        }
+        if (user.emailVerified) {
+          return res.status(409).json({
+            status: 'failed',
+            message: 'Your account had already been verified'
+          });
+        }
+
+        const emailToken = jwt.sign({ id: user.id }, secret, { expiresIn: '1h' });
+        const url = `http://localhost:3000/api/users/confirmation/${emailToken}`;
```
Comment (aknwosu) [inline server/controllers/emailVerificationController.js:115]: have this in your env file

File: server/controllers/emailVerificationController.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        sgMail.setApiKey(process.env.SENDGRID_API_KEY);
+        const msg = {
+          to: email,
+          from: 'noreply@authorshaven.com',
+          subject: 'Authors Haven Email Verification',
```
Comment (aknwosu) [inline server/controllers/emailVerificationController.js:120]: Authors' Haven Email Verification. apostrophe missing

File: server/controllers/users.js [hunk 5]
```diff
@@ -0,0 +1,216 @@
+import bcrypt from 'bcrypt';
+
+import db from '../models';
+import EmailVerificationController from '../controllers/emailVerificationController';
+import isValidNumber from '../utils/is_valid_number';
+import UserValidation from '../validation/users';
+import trimInput from '../utils/trim_input';
+
+const { User } = db;
```
Comment (aknwosu) [inline server/controllers/users.js:9]: import { User } from '../models'

File: server/controllers/users.js [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+            errors: {
+              email: 'User not found'
+            }
+          });
+        }
+        bcrypt.compare(password, user.hash).then((isMatch) => {
+          if (isMatch) {
+            return res.status(200).json({
+              message: 'Login successful',
+              user,
```
Comment (aknwosu) [inline server/controllers/users.js:96]: check the response object from others and make this correspond

File: server/controllers/users.js [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+              lastName: trimInput(lastName) || user.lastName,
+              username: trimInput(username) || user.username,
+              bio: trimInput(bio) || user.bio,
+              twitter: twitter || user.twitter,
+              linkedin: linkedin || user.linkedin,
+              facebook: facebook || user.facebook,
+              image: image || user.image,
+            }).then((updatedUser) => {
+              const { dataValues } = updatedUser;
+              return res.status(200).json({ dataValues });
```
Comment (aknwosu) [inline server/controllers/users.js:209]: correct all your response objects and standardize them

Comment (aknwosu) [inline server/controllers/users.js:209]: really? anything as the response??

File: server/controllers/emailVerificationController.js [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+   * @param  {function} id of the new user
+   * @returns {object} The body of the resposne message
+   * @memberof EmailVerificationController
+   */
+  static sendVerificationEmail(user) {
+    try {
+      const { id, email } = user;
+      const emailToken = jwt.sign({ email, id }, secret, { expiresIn: '1h' }); 
+
+      const url = `http://localhost:3000/api/users/confirmation/${emailToken}`;
```
Comment (aknwosu) [inline server/controllers/emailVerificationController.js:31]: if you hardcode the url here, what happens in production?

**Review by aknwosu — CHANGES_REQUESTED** [review]

**Comment by aknwosu** [comment]
I don't expect to still have to ask anyone to fix hound issues at this point

**Pull request closed without merging**